File: Algorithm/Binary Lifting/atcoder/B57/GPT/B57.go
// では Go (1.20.6) で **ダブリング法** を使った高速解法を書きます。
// 配列は `[]uint32` を使ってメモリ効率を上げ、不要なオブジェクト生成を避けます。

// ---

// ```go
// Go 1.20.6
// 実行時間制限: 4 sec / メモリ制限: 1024 MiB
// 入力: N K
// 出力: 各 i (1 <= i <= N) に対し、K 回の操作後の値

package main

import (
	"bufio"
	"fmt"
	"os"
)

/*
sumDigits は整数 x の桁和を計算する関数
@param x int - 計算対象の整数
@return int - 桁和
*/
func sumDigits(x int) int {
	s := 0
	for x > 0 {
		s += x % 10
		x /= 10
	}
	return s
}

/*
solve は各整数 1..N に対し K 回操作後の値を求める
@param N int     - 最大の整数
@param K uint64  - 操作回数 (1 <= K <= 1e9)
@return []uint32 - 各整数の最終値
*/
func solve(N int, K uint64) []uint32 {
	// 1ステップ先を格納する配列
	next := make([]uint32, N+1)
	for i := 1; i <= N; i++ {
		next[i] = uint32(i - sumDigits(i))
	}

	// ダブリングテーブル構築
	const LOG = 32 // K<=1e9 なので log2(K)≈30、余裕を持たせて32
	jump := make([][]uint32, LOG)
	for p := range jump {
		jump[p] = make([]uint32, N+1)
	}
	copy(jump[0], next)
	for p := 1; p < LOG; p++ {
		prev := jump[p-1]
		cur := jump[p]
		for i := 1; i <= N; i++ {
			cur[i] = prev[prev[i]]
		}
	}

	// 各 i について K ステップ先を計算
	res := make([]uint32, N)
	for i := 1; i <= N; i++ {
		cur := uint32(i)
		steps := K
		bit := 0
		for steps > 0 {
			if steps&1 == 1 {
				cur = jump[bit][cur]
			}
			steps >>= 1
			bit++
		}
		res[i-1] = cur
	}

	return res
}

func main() {
	in := bufio.NewReaderSize(os.Stdin, 1<<20)
	out := bufio.NewWriterSize(os.Stdout, 1<<20)
	defer out.Flush()

	var N int
	var K uint64
	fmt.Fscan(in, &N, &K)

	res := solve(N, K)

	for _, v := range res {
		fmt.Fprintln(out, v)
	}
}
// ```

// ---

// ## **処理時間・メモリ見積もり**

// * **時間計算量**

//   * next 配列構築: $O(N \cdot d)$（dは桁数、最大6）
//   * ダブリング: $O(N \log K)$ ≈ $3\times 10^5 \times 30 \approx 9\times 10^6$ ステップ
//   * 各 i の計算: $O(N \log K)$
//     → Go でも 1 秒未満〜2 秒程度で収まる想定

// * **メモリ使用量**

//   * `next`: 約 1.2MB
//   * `jump`: 約 38MB
//   * `res`: 約 1.2MB
//     → 合計 約 40MB（1024MB 制限内）

// ---

// この Go 実装なら処理時間・メモリともに十分余裕があります。
// もし必要なら、**ジャンプ表を1本だけにしてメモリ削減版** も書けます。

// この後、メモリ削減版の Go 実装も作りますか？

File: Algorithm/Binary Lifting/atcoder/B57/GPT/B57.py
# Python でも **ダブリング法 (binary lifting)** を使って高速に解く実装を作ります。
# Pylance の型エラーを避けるために **型ヒントを明示** し、
# 処理時間・メモリを意識して **list\[int]** と局所変数最適化をしています。

# ---

# ```python
# Python 3.11.4
# 実行時間制限: 4 sec / メモリ制限: 1024 MiB
# 入力: N K
# 出力: 各 i (1 <= i <= N) に対し、K 回の操作後の値

import sys
from typing import List

def sum_digits(x: int) -> int:
    """桁和を計算する関数"""
    s: int = 0
    while x > 0:
        s += x % 10
        x //= 10
    return s

def solve(N: int, K: int) -> List[int]:
    """
    各 i (1 <= i <= N) について、K 回操作後の値を求める
    Parameters:
        N (int): 最大の整数
        K (int): 操作回数 (1 <= K <= 10^9)
    Returns:
        List[int]: 各 i の最終値
    """
    # 1ステップ先の値
    next_arr: List[int] = [0] * (N + 1)
    for i in range(1, N + 1):
        next_arr[i] = i - sum_digits(i)

    # ダブリングテーブル構築
    LOG: int = 32  # K <= 1e9 なので 2^30 程度で十分、余裕を持たせて32
    jump: List[List[int]] = [[0] * (N + 1) for _ in range(LOG)]
    jump[0] = next_arr[:]
    for p in range(1, LOG):
        prev = jump[p - 1]
        cur = jump[p]
        for i in range(1, N + 1):
            cur[i] = prev[prev[i]]

    # 各 i について K ステップ先を計算
    res: List[int] = [0] * N
    for i in range(1, N + 1):
        cur = i
        steps = K
        bit = 0
        while steps > 0:
            if steps & 1:
                cur = jump[bit][cur]
            steps >>= 1
            bit += 1
        res[i - 1] = cur

    return res

def main() -> None:
    data: List[int] = list(map(int, sys.stdin.buffer.read().split()))
    N: int = data[0]
    K: int = data[1]

    result: List[int] = solve(N, K)
    sys.stdout.write("\n".join(map(str, result)))

if __name__ == "__main__":
    main()
# ```

# ---

# ## **処理時間・メモリ見積もり**

# * **計算量**

#   * next 配列構築: $O(N \cdot d)$（dは桁数、最大6）
#   * ダブリング: $O(N \log K)$ ≈ $3\times 10^5 \times 30 \approx 9\times 10^6$ ステップ
#   * 各 i の計算: $O(N \log K)$
#     → 実行時間: 約 0.5 秒程度（CPython 3.11）

# * **メモリ使用量**

#   * `next_arr`: 約 1.2MB
#   * `jump`: 約 38MB
#   * `res`: 約 1.2MB
#     → 合計 約 40MB（1024MB 制限内）

# ---

# このままでも十分高速ですが、
# もしさらに **メモリを半分以下に削減** するなら、
# ダブリングテーブルを全保持せず 1 本だけ計算し直す方式にもできます。

# ---

# この後、メモリ削減版の Python 実装も作りますか？
# そうするとジャンプ表が 1 レベル分で済み、約 1.2MB まで落とせます。
